src/PluginForm/Bluesnap/PaymentMethodAddForm.php

```diff
@@ -0,0 +1,176 @@
+    ];
+    // Hidden fields which will be populated by the js once a card is
+    // successfully submitted to BlueSnap. This will be used for the card
+    // details for Anonymous users when creating a payment method.
+    $element['bluesnap_cc_type'] = [
+      '#type' => 'hidden',
+      '#attributes' => [
+        'id' => 'bluesnap-cc-type',
+      ],
+    ];
+    $element['bluesnap_cc_last_4'] = [
+      '#type' => 'hidden',
+      '#attributes' => [
+        'id' => 'bluesnap-cc-last-4',
+      ],
+    ];
+    $element['bluesnap_cc_expiry'] = [
+      '#type' => 'hidden',
+      '#attributes' => [
+        'id' => 'bluesnap-cc-expiry',
+      ],
+    ];
+
+    // The credit card fields necessary for BlueSnap.
+    $element['card_number'] = [
+      '#type' => 'item',
+      '#title' => t('Card number'),
+      '#required' => TRUE,
+      '#validated' => TRUE,
+      '#markup' => '<div class="form-control" id="card-number" data-bluesnap="ccn"></div>',
+    ];
+
+    $element['expiration'] = [
+      '#type' => 'item',
+      '#title' => t('Expiration date'),
+      '#required' => TRUE,
+      '#validated' => TRUE,
+      '#markup' => '<div class="form-control" id="exp-date" data-bluesnap="exp"></div>',
+    ];
+
+    $element['security_code'] = [
+      '#type' => 'item',
+      '#title' => t('CVV'),
+      '#required' => TRUE,
+      '#validated' => TRUE,
+      '#markup' => '<div class="form-control" id="cvv" data-bluesnap="cvv"></div>',
+    ];
+
+    // To display validation errors.
+    $element['payment_errors'] = [
+      '#type' => 'markup',
+      '#markup' => '<div id="payment-errors"></div>',
+      '#weight' => -200,
+    ];
+
+    return $element;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  protected function validateCreditCardForm(array &$element, FormStateInterface $form_state) {
+    // The JS library performs its own validation.
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function submitCreditCardForm(array $element, FormStateInterface $form_state) {
+    // The payment gateway plugin will process the submitted payment details.
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function buildConfigurationForm(array $form, FormStateInterface $form_state) {
+    $form = parent::buildConfigurationForm($form, $form_state);
+
+    // Add the bluesnap attribute to the postal code field.
+    $form['billing_information']['address']['widget'][0]['address_line1']['#attributes']['data-bluesnap'] = 'address_line1';
+    $form['billing_information']['address']['widget'][0]['address_line2']['#attributes']['data-bluesnap'] = 'address_line2';
+    $form['billing_information']['address']['widget'][0]['locality']['#attributes']['data-bluesnap'] = 'address_city';
+    $form['billing_information']['address']['widget'][0]['postal_code']['#attributes']['data-bluesnap'] = 'address_zip';
+    $form['billing_information']['address']['widget'][0]['country_code']['#attributes']['data-bluesnap'] = 'address_country';
+
+    return $form;
+  }
```

Comment:
**CHANGE REQUEST**

*Readability/Coding standards*

```
// Add the bluesnap attribute to address form elements.
$element = &$form['billing_information']['address']['widget'][0];
$element['address_line1']['#attributes']['data-bluesnap'] = 'address_line1';
// etc.
```
Corrected the comment as well as we're adding the attributes to all address form elements, not just the postal code.
```